### PR TITLE
Add forbidden api checker

### PR DESCRIPTION
**Pull request by BLoHny**
relates https://github.com/mfvanek/pg-index-health/issues/195

### CODE REVIEW

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -302,10 +302,5 @@
         <module name="CommentsIndentation">
             <property name="tokens" value="SINGLE_LINE_COMMENT, BLOCK_COMMENT_BEGIN"/>
         </module>
-        <module name="IllegalMethodCall">
```
Comment (mfvanek) [inline config/checkstyle/checkstyle.xml:305]: It looks like we can get rid of `checkstyle("com.thomasjensen.checkstyle.addons:checkstyle-addons:7.0.1")`

File: build.gradle.kts [hunk 2]
```diff
@@ -68,3 +69,23 @@ tasks.named<DependencyUpdatesTask>("dependencyUpdates").configure {
...
     }
 }
+
+forbiddenApis {
+    bundledSignatures = setOf("jdk-deprecated", "jdk-non-portable")
+    signaturesFiles = files("/config/forbidden-apis/forbidden-apis.txt")
+    ignoreFailures = false
+}
+
+subprojects {
```
Comment (mfvanek) [inline build.gradle.kts:79]: We do not use `subprojects` any more. Please move it to the `buildSrc`.
See `buildSrc/src/main/kotlin/pg-index-health.pitest.gradle.kts` for example

**Review by mfvanek — CHANGES_REQUESTED** [review]

File: build.gradle.kts [hunk 3]
```diff
@@ -68,3 +69,23 @@ tasks.named<DependencyUpdatesTask>("dependencyUpdates").configure {
         isNonStable(candidate.version)
     }
 }
+
+forbiddenApis {
+    bundledSignatures = setOf("jdk-deprecated", "jdk-non-portable")
```
Comment (mfvanek) [inline build.gradle.kts:74]: Let's try to use all bundled signatures https://github.com/policeman-tools/forbidden-apis/wiki/BundledSignatures

Comment (BLoHny) [reply build.gradle.kts:74]: When adding build signatures such as Unsafe or Refelection in existing changes, errors occurred in existing code such as String.format();
After searching about this, I found this issue https://github.com/policeman-tools/forbidden-apis/issues/119
I'm not sure if I understand this issue correctly.
Based on the four build signatures in the example in the issue of this pull request, I added only two build signatures that did not cause errors in the existing code.

How should we proceed?

Comment (mfvanek) [reply build.gradle.kts:74]: > How should we proceed?

Use
```
bundledSignatures = setOf("jdk-unsafe", "jdk-deprecated", "jdk-internal", "jdk-non-portable", "jdk-system-out", "jdk-reflection")
```
and rewrite `String.format` usages

**Comment by BLoHny** [comment]
The build runs fine locally. Could this be a CI error caused by changing setAccessible in DatabaseCheckTest?

File: pg-index-health-core/src/main/java/io/github/mfvanek/pg/common/maintenance/Diagnostic.java [hunk 4]
```diff
@@ -24,25 +24,25 @@
...
-    COLUMNS_WITH_JSON_TYPE(ExecutionTopology.ON_PRIMARY, "columns_with_json_type.sql", QueryExecutors::executeQueryWithSchema),
-    COLUMNS_WITH_SERIAL_TYPES(ExecutionTopology.ON_PRIMARY, "non_primary_key_columns_with_serial_types.sql", QueryExecutors::executeQueryWithSchema),
-    FUNCTIONS_WITHOUT_DESCRIPTION(ExecutionTopology.ON_PRIMARY, "functions_without_description.sql", QueryExecutors::executeQueryWithSchema),
-    INDEXES_WITH_BOOLEAN(ExecutionTopology.ON_PRIMARY, "indexes_with_boolean.sql", QueryExecutors::executeQueryWithSchema),
-    NOT_VALID_CONSTRAINTS(ExecutionTopology.ON_PRIMARY, "check_not_valid_constraints.sql", QueryExecutors::executeQueryWithSchema),
-    BTREE_INDEXES_ON_ARRAY_COLUMNS(ExecutionTopology.ON_PRIMARY, "btree_indexes_on_array_columns.sql", QueryExecutors::executeQueryWithSchema),
-    SEQUENCE_OVERFLOW(ExecutionTopology.ON_PRIMARY, "sequence_overflow.sql", QueryExecutors::executeQueryWithRemainingPercentageThreshold);
+    BLOATED_INDEXES(ExecutionTopology.ON_PRIMARY, "sql/bloated_indexes.sql", QueryExecutors::executeQueryWithBloatThreshold),
```
Comment (mfvanek) [inline pg-index-health-core/src/main/java/io/github/mfvanek/pg/common/maintenance/Diagnostic.java:27]: Revert these changes

File: pg-index-health-logger/src/main/java/io/github/mfvanek/pg/common/maintenance/DatabaseChecks.java [hunk 5]
```diff
@@ -74,11 +75,15 @@ public DatabaseChecks(@Nonnull final HighAvailabilityPgConnection haPgConnection
...
-            throw new IllegalStateException(String.format("Check for diagnostic %s not found", diagnostic));
+            throw new IllegalStateException(String.format(Locale.ROOT, "Check for diagnostic %s not found", diagnostic));
         }
         if (!type.isAssignableFrom(check.getType())) {
             throw new IllegalArgumentException("Illegal type: " + type);
         }
         return (DatabaseCheckOnCluster<T>) check;
     }
+
+    public void clearChecks() {
```
Comment (mfvanek) [inline pg-index-health-logger/src/main/java/io/github/mfvanek/pg/common/maintenance/DatabaseChecks.java:86]: Please remove this method

File: buildSrc/src/main/kotlin/pg-index-health.java-conventions.gradle.kts [hunk 6]
```diff
@@ -95,3 +96,9 @@ spotbugs {
     reportLevel.set(Confidence.LOW)
     excludeFilter.set(file("../config/spotbugs/exclude.xml"))
 }
+
+forbiddenApis {
```
Comment (mfvanek) [inline buildSrc/src/main/kotlin/pg-index-health.java-conventions.gradle.kts:100]: It should be only in one place (in buildSrc/src/main/kotlin/pg-index-health.forbidden-apis.gradle.kts)

File: buildSrc/src/main/kotlin/pg-index-health.java-conventions.gradle.kts [hunk 7]
```diff
@@ -14,10 +14,11 @@ plugins {
     id("java-test-fixtures")
     id("net.ltgt.errorprone")
     id("org.gradle.test-retry")
+    id("de.thetaphi.forbiddenapis")
 }
 
 dependencies {
-    checkstyle("com.thomasjensen.checkstyle.addons:checkstyle-addons:7.0.1")
+    implementation("de.thetaphi:forbiddenapis:3.7")
```
Comment (mfvanek) [inline buildSrc/src/main/kotlin/pg-index-health.java-conventions.gradle.kts:21]: Please remove it

File: buildSrc/src/main/kotlin/pg-index-health.java-conventions.gradle.kts [hunk 8]
```diff
@@ -36,7 +37,7 @@ tasks.withType<JavaCompile>().configureEach {
 
 tasks {
     test {
-        dependsOn(checkstyleMain, checkstyleTest, pmdMain, pmdTest, spotbugsMain, spotbugsTest)
+        dependsOn(checkstyleMain, checkstyleTest, pmdMain, pmdTest, spotbugsMain, spotbugsTest, forbiddenApisMain, forbiddenApisTest)
```
Comment (mfvanek) [inline buildSrc/src/main/kotlin/pg-index-health.java-conventions.gradle.kts:40]: Move `(forbiddenApisMain, forbiddenApisTest)` to the buildSrc/src/main/kotlin/pg-index-health.forbidden-apis.gradle.kts

File: build.gradle.kts [hunk 9]
```diff
@@ -5,6 +5,7 @@ plugins {
     id("jacoco-report-aggregation")
     id("org.sonarqube")
     id("com.github.ben-manes.versions") version "0.51.0"
+    id("de.thetaphi.forbiddenapis")
```
Comment (mfvanek) [inline build.gradle.kts:8]: I think we don't need it here

File: pg-index-health-core/src/test/java/io/github/mfvanek/pg/checks/host/ColumnsWithSerialTypesCheckOnHostTest.java [hunk 10]
```diff
@@ -44,9 +46,9 @@ void onDatabaseWithThem(final String schemaName) {
                 .hasSize(2)
                 .containsExactly(
                     ColumnWithSerialType.ofBigSerial(
-                        Column.ofNotNull(ctx.enrichWithSchema("bad_accounts"), "real_account_id"), String.format("%s.bad_accounts_real_account_id_seq", schemaName)),
+                        Column.ofNotNull(ctx.enrichWithSchema("bad_accounts"), "real_account_id"), String.format(Locale.ROOT, "%s.bad_accounts_real_account_id_seq", schemaName)),
```
Comment (mfvanek) [inline pg-index-health-core/src/test/java/io/github/mfvanek/pg/checks/host/ColumnsWithSerialTypesCheckOnHostTest.java:49]: It looks like we can add a new method `enrichSequenceWithSchema` to the `PgContext`.
It can be like `ctx.enrichSequenceWithSchema("bad_accounts_real_client_id_seq")`

**Review by mfvanek — CHANGES_REQUESTED** [review]

File: pg-index-health-core/src/test/java/io/github/mfvanek/pg/checks/host/ColumnsWithSerialTypesCheckOnHostTest.java (hunk 10 above)
Comment (mfvanek) [reply pg-index-health-core/src/test/java/io/github/mfvanek/pg/checks/host/ColumnsWithSerialTypesCheckOnHostTest.java:49]: I've added a new issue for that https://github.com/mfvanek/pg-index-health/issues/422

File: pg-index-health-logger/src/main/java/io/github/mfvanek/pg/common/maintenance/DatabaseChecks.java (hunk 5 above)
Comment (BLoHny) [reply pg-index-health-logger/src/main/java/io/github/mfvanek/pg/common/maintenance/DatabaseChecks.java:86]: The use of setAccessible in io.github.mfvanek.pg.support.TestUtils and io.github.mfvanek.pg.common.maintenance.DatabaseChecksTest causes an error in the foribdden-api check due to java-refelection. How should I handle this?

https://github.com/policeman-tools/forbidden-apis/issues/198
Currently forbidden-api doesn't seem to support excluding specific classes.

Comment (mfvanek) [reply pg-index-health-logger/src/main/java/io/github/mfvanek/pg/common/maintenance/DatabaseChecks.java:86]: > Should we try SuppressForbidden?

Yes, try to use https://github.com/policeman-tools/forbidden-apis/blob/main/src/main/java/de/thetaphi/forbiddenapis/SuppressForbidden.java

**Review by mfvanek — APPROVED** [review]
